File: java/src/main/java/codingdojo/Fizzbuzz6.java
package codingdojo;

import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Fizzbuzz6 {
    public static void fizzbuzz() {
        new Fizzbuzz6().calculateFizzbuzz().forEach(System.out::println);
    }

    private static List<FizzBuzzFactor> factors = Arrays.asList(
            new FizzBuzzFactor("Fizz", 3),
            new FizzBuzzFactor("Buzz", 5)
    );
    private final FizzbuzzRange range;

    public Fizzbuzz6() {
        this.range = new FizzbuzzRange(1, 101);
    }

    Stream<FizzbuzzProvider> calculateFizzbuzz() {
        return this.range.stream().map( (n) -> new FizzbuzzCalculator().convert(n));
    }

    static class FizzbuzzCalculator {
        public FizzbuzzProvider convert(int n) {
            StringBuilder result = new StringBuilder();
            for (Fizzbuzz6.FizzBuzzFactor factor : factors) {
                if (factor.predicate.test(n)) {
                    result.append(factor.name);
                }
            }
            return new FizzbuzzProvider(result.toString(), n);
        }
    }

    static class FizzbuzzProvider {
        private int n;
        private String provided;

        public  FizzbuzzProvider(String provided, int n) {
            this.provided = provided;
            this.n = n;
        }
        public String provide() {
            return "".equals(provided) ? "" + this.n : provided;
        }
        @Override
        public String toString() {
            return provide();
        }
    }

    static class FizzbuzzRange {

        private final int from;
        private final int to;

        public FizzbuzzRange(int from, int to) {
            this.from = from;
            this.to = to;
        }

        public Stream<Integer> stream() {
            return IntStream.range(this.from, this.to).boxed();
        }
    }

    static class FizzBuzzFactor {
        public String name;
        public Predicate<Integer> predicate;
        public FizzBuzzFactor(String name, int factor) {
            this.name = name;
            this.predicate = n -> n % factor == 0;
        }

    }
}

File: kotlin/src/main/kotlin/codingdojo/Fizzbuzz6.kt
package codingdojo

class Fizzbuzz6 {
    companion object {
        fun fizzbuzz() {
            val fizzbuzz6 = Fizzbuzz6();
            fizzbuzz6.calculateFizzbuzz().forEach(System.out::println);
        }
    }

    val range = Range(1, 100);
    val factors = listOf(Factor("Fizz", 3), Factor("Buzz", 5));

    fun calculateFizzbuzz(): List<Provider> {
        val calculator = FizzbuzzCalculator(factors);
        return this.range.stream().map{i -> calculator.convert(i)};
    }

    class FizzbuzzCalculator(val factors: List<Factor>) {

        fun convert(n: Int): Provider {
            val result = StringBuilder();
            for (factor in factors) {
                if (factor.predicate(n)) {
                    result.append(factor.name);
                }
            }
            return Provider(n, result.toString());
        }
    }

    class Provider(val n: Int, val provided: String) {

        fun provide(): String {
            if ("".equals(provided))
                return "$n"
            return provided;
        }

        override fun toString(): String {
            return provide();
        }
    }

    class Range(val from: Int, val to: Int) {
        fun stream(): IntRange {
            return (this.from..this.to)
        }
    }

    class Factor(val name: String, val factor: Int) {
        val predicate =  {n: Int -> n % factor == 0}
    }
}
